rattan-core/src/cells/token_bucket.rs

```diff
@@ -0,0 +1,968 @@
+use std::cmp::{max, min};
+use std::fmt::Debug;
+use std::sync::atomic::AtomicI32;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+use tokio::time::{Duration, Instant};
+use tracing::debug;
+
+use super::bandwidth::queue::{DropTailQueue, DropTailQueueConfig};
+use super::bandwidth::LARGE_DURATION;
+use super::{ControlInterface, Egress, Ingress};
+
+pub const LARGE_RATE_LIMIT: u64 = 1073741824;
+pub const LARGE_BUFFER: usize = 1073741824;
+
+// transfer byte into time
+fn length_to_time(length: usize, rate: u64) -> Duration {
+    if length == 0 {
+        Duration::from_secs(0)
+    } else if rate == 0 {
+        LARGE_DURATION
+    } else {
+        Duration::from_secs_f64(length as f64 / rate as f64)
+    }
+}
+
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[derive(Debug, Clone, Copy)]
+pub struct TokenBucket {
+    buffer: usize,
+    rate: u64,
+    tokens: i64,
+}
+
+impl Default for TokenBucket {
+    fn default() -> Self {
+        Self {
+            buffer: LARGE_BUFFER,
+            rate: LARGE_RATE_LIMIT,
+            tokens: 0,
+        }
+    }
+}
+
+impl TokenBucket {
+    pub fn new(buffer: usize, rate: u64) -> Self {
+        Self {
+            buffer,    // byte
+            rate,      // byte/s
+            tokens: 0, // ns
+        }
+    }
+}
+
+pub struct TokenBucketCellIngress<P>
+where
+    P: Packet,
+{
+    ingress: mpsc::UnboundedSender<P>,
+}
+
+impl<P> Clone for TokenBucketCellIngress<P>
+where
+    P: Packet,
+{
+    fn clone(&self) -> Self {
+        Self {
+            ingress: self.ingress.clone(),
+        }
+    }
+}
+
+impl<P> Ingress<P> for TokenBucketCellIngress<P>
+where
+    P: Packet + Send,
+{
+    fn enqueue(&self, mut packet: P) -> Result<(), Error> {
+        packet.set_timestamp(Instant::now());
+        self.ingress
+            .send(packet)
+            .map_err(|_| Error::ChannelError("Data channel is closed.".to_string()))?;
+        Ok(())
+    }
+}
+
+pub struct TokenBucketCellEgress<P>
+where
+    P: Packet,
+{
+    egress: mpsc::UnboundedReceiver<P>,
+    max_size: usize,
+    token_bucket: TokenBucket,
+    peak_token_bucket: TokenBucket,
+    packet_queue: DropTailQueue<P>,
+    next_check: Instant,
+    last_check: Instant,
+    config_rx: mpsc::UnboundedReceiver<TokenBucketCellConfig>,
+    timer: Timer,
+    state: AtomicI32,
+}
+
+impl<P> TokenBucketCellEgress<P>
+where
+    P: Packet + Send + Sync,
+{
+    fn get_new_tb(&self, new_rate: Option<u64>, new_burst: Option<usize>) -> TokenBucket {
+        let mut new_tb = match (new_rate, new_burst) {
+            (None, None) => self.token_bucket,
+            (Some(rate), None) => TokenBucket::new(self.token_bucket.buffer, rate),
+            (None, Some(burst)) => TokenBucket::new(burst, self.token_bucket.rate),
+            (Some(rate), Some(burst)) => TokenBucket::new(burst, rate),
+        };
+        new_tb.tokens = length_to_time(new_tb.buffer, new_tb.rate).as_nanos() as i64;
+        new_tb
+    }
```

Comment:
How should I deal with the situation where the burst equals 0?

Reply:
Reject such configuration (return error) if you think it is an unreasonable parameter (I mean, if you think tbf will function abnormally with burst equal to 0). Or if you think making TBF not store and send any packet is a kind of choice, you can just ignore my comment.

PS: I suppose this is an abnormal case, but we can discuss it if you have different opinions.

Reply:
In my implementation the tbf will drop all packets when burst is equal to 0, because it will drop any packet the size of which is larger than `max_size` (the smaller one between burst and minburst). Is this an abnormal case?

Reply:
This is the case I want to argue with you previously: when should we drop the packets. If the packet enter our inner queue but it cannot be sent since the burst is smaller than its packet size. And at some point, the configuration of our TBF is changed, where the burst size is large than this packet's size. Now it can be sent. So it does matter when you drop the packet. It will influence exactly which packets can be sent after the configuration changes.

I can accept the case with 0 burst to send no packet, but you should still pay attention to when and where to drop the packets. This will, in fact, lead to different drop mechanisms (drop-tail or drop-head).

Reply:
But our tbf will only check whether the front packet of the queue can be sent or not. If a packet is larger than max_size, it will never be sent since it's too large compared to our burst and minburst, and our tbf will be blocked at this large packet. In tc-tbf, these packets are dropped before enqueueing. So I do the same thing. If max_size is changed, I drop those packet already in the queue but larger than new max_size when the configuration changes. I think it's the same as not allowing them to enqueue from the beginning, and indeed, the modified bucket cannot send these packets either.

Reply:
> If max_size is changed, I drop those packet already in the queue but larger than new max_size when the configuration changes. I think it's the same as not allowing them to enqueue from the beginning, and indeed, the modified bucket cannot send these packets either.

I'm still trying to figure it out. Please explain the situation further: suppose the size of our packet is 1500, and the current max_size is 1000. Will you drop this packet before enqueuing it into our internal queue? After 1 second, the max_size is somehow changed to 1600, and if that packet is enqueued and thus waiting in our internal queue, it can be sent now. It will have opportunity to not block the operation. This is different with kernel's tc-tbf, as our max_size could be enlarged to allow some packets to pass. I think these two cases are different.

Reply:
I know what you mean exactly. So how should we deal with this situation? We can never assume the configuration will change so that a packet larger than max_size will be able to be sent. Also we can neither know when the configuration will be changed. So maybe not enqueueing these large packets is already a relatively good method I think.

Reply:
Ok, it is fine that we have such limitations. Then you should clearly state such things in detail in the document comments of the TBF struct and its configuration struct. Then we should disallow the situation of zero burst, as tc-tbf will throw error if you set burst/miniburst to zero.

Besides, as what illustrated in the tc-tbf example:

```bash
tc qdisc add dev eth0 handle 10: root tbf rate 0.5mbit \
         burst 5kb latency 70ms peakrate 1mbit       \
         minburst 1540
```

We could use bandwidth (bit per second) as the unit of our rate, at least for the interactive interfaces (config). Similar to BW cell, you can use crate `bandwidth` and `human-bandwidth`. Also, to configure burst and miniburst, we can use the crate `humansize`

Reply:
Offer you one more reference on the configuration value check (as I find `rate=0` is also disallowed): [iprout2/iproute2/tc/q-tbf.c](https://github.com/iproute2/iproute2/blob/413cf4f03a9b6a219c94b86f41d67992b0a14b82/tc/q_tbf.c#L174-L192)

Reply:
I agree with your design choice. We don't need to provide the capability to revert the tbf to an unset state. And we can give the TBF config default value (large value), and `rate` and `burst` should be enforced to give together